pubsub/averager.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub
+
+import (
+	"sync"
+	"time"
+)
+
+// An averager keeps track of an average value over a time interval.
+// It does so by tracking the average over several sub-intervals.
+type averager struct {
+	mu             sync.Mutex
+	buckets        []bucket // always nBuckets of these, last is most recent
+	bucketInterval time.Duration
+	end            time.Time // latest time we can currently record (end of the last bucket's interval)
+}
+
+type bucket struct {
+	count float64 // number of points in the bucket
+	avg   float64 // average value of the bucket
```

Comment:
Are you keeping `avg` instead of `total` to avoid overflow? You seem to always multiply `avg` by `count`. If so, add a comment?

Reply:
Just being stupid. Rewrote.